x/feedistribution/keeper/allocation.go

```diff
@@ -0,0 +1,171 @@
+	// if the rewards do not exist it's fine, we will just add to zero.
+	// allocate share tokens to all stakers of this operator.
+	operatorAccAddress := sdk.AccAddress(valBz)
+	k.AllocateTokensToStakers(ctx, operatorAccAddress, shared, feePool)
+
+	// update outstanding rewards
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeRewards,
+		sdk.NewAttribute(sdk.AttributeKeyAmount, commission.String()),
+		sdk.NewAttribute(types.AttributeKeyValidator, val.GetOperator().String()),
+	))
+
+	// ValidatorOutstandingRewards is the rewards of a validator address.
+	outstanding := k.GetValidatorOutstandingRewards(ctx, valBz)
+	outstanding.Rewards = outstanding.Rewards.Add(tokens...)
+	k.SetValidatorOutstandingRewards(ctx, valBz, outstanding)
+	logger.Info("Allocate tokens to validator successfully", "allocated amount is", outstanding.Rewards.String())
+}
+
+func (k Keeper) AllocateTokensToStakers(ctx sdk.Context, operatorAddress sdk.AccAddress, rewardToAllStakers sdk.DecCoins, feePool *types.FeePool) {
+	logger := k.Logger()
+	logger.Info("Allocate to stakers of operatorAddress:", operatorAddress.String())
+	avsList, err := k.StakingKeeper.GetOptedInAVSForOperator(ctx, operatorAddress.String())
+	if err != nil {
+		logger.Debug("avs address lists not found; skipping")
+		return
+	}
+	stakersPowerMap, curTotalStakersPowers := make(map[string]math.LegacyDec), math.LegacyNewDec(0)
+	globalStakerAddressList := make([]string, 0)
+	for _, avsAddress := range avsList {
+		avsAssets, err := k.StakingKeeper.GetAVSSupportedAssets(ctx, avsAddress)
+		if err != nil {
+			logger.Debug("avs address lists not found; skipping")
+			continue
+		}
+		for assetID := range avsAssets {
+			stakerList, err := k.StakingKeeper.GetStakersByOperator(ctx, operatorAddress.String(), assetID)
+			if err != nil {
+				logger.Debug("staker lists not found; skipping")
+				continue
+			}
+			for _, staker := range stakerList.Stakers {
+				if curStakerPower, err := k.StakingKeeper.CalculateUSDValueForStaker(ctx, staker, avsAddress, operatorAddress.Bytes()); err != nil {
+					logger.Error("curStakerPower error", err)
+				} else {
+					stakersPowerMap[staker] = curStakerPower
+					globalStakerAddressList = append(globalStakerAddressList, staker)
+					curTotalStakersPowers = curTotalStakersPowers.Add(curStakerPower)
+				}
+			}
+		}
+	}
+	sort.Slice(globalStakerAddressList, func(i, j int) bool {
+		return stakersPowerMap[globalStakerAddressList[i]].GT(stakersPowerMap[globalStakerAddressList[j]])
+	})
+
+	if curTotalStakersPowers.IsPositive() {
+		// allocate to stakers in voting power descending order if the curTotalStakersPower is positive
+		for _, staker := range globalStakerAddressList {
+			stakerPower := stakersPowerMap[staker]
+			powerFraction := stakerPower.QuoTruncate(curTotalStakersPowers)
+			rewardToSingleStaker := rewardToAllStakers.MulDecTruncate(powerFraction)
+			if rw := rewardToAllStakers.Sub(rewardToSingleStaker); rw.IsValid() {
```

Comment:
the logic here wrong, rewardToAllStakers value should **NOT** be changed, another variable like remaining should be defined and initialize with rewardToAllStakers, then for each staker, after calculating the reward for the staker, update the remaining reward value to check if all rewards are distributed.